### PR TITLE
[SPARK-5579][SQL][DataFrame] Support for project/filter using SQL expressions

### PR DESCRIPTION
**Pull request by rxin**
``` scala
df.selectExpr("abs(colA)", "colB")
df.filter("age > 21")
```

### CODE REVIEW

File: python/pyspark/sql.py [hunk 1]
```diff
@@ -2126,10 +2126,9 @@ def sort(self, *cols):
         """
         if not cols:
             raise ValueError("should sort by at least one column")
-        jcols = ListConverter().convert([_to_java_column(c) for c in cols[1:]],
+        jcols = ListConverter().convert([_to_java_column(c) for c in cols],
```
Comment (rxin) [inline python/pyspark/sql.py]: @davies take a look at the Python changes.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala [hunk 2]
```diff
@@ -179,10 +179,20 @@ private[sql] class DataFrameImpl protected[sql](
     select((col +: cols).map(Column(_)) :_*)
   }
 
+  override def selectExpr(exprs: String*): DataFrame = {
```
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala]: I think this one could be merged into select(), column is also a valid expression

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala]: not if it has space ... it will just fail

**Comment by davies** [comment]
This select() and filter() in Python do not support yet

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala (hunk 2 above)
Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala]: It should work in these cases with this implementation.
```
select('a', '`the name`', 'a + 1', 'min(b) * 3')
```

Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrameImpl.scala]: yea - but asking users to wrap a column name in backticks in strings is fairly annoying.

**Comment by rxin** [comment]
We can discuss more offline. For now let's keep this separate, otherwise it can be fairly annoying to use column names that contain space or column names that contain any SQL keywords.

**Pull request closed without merging**